tests/Agency/StorePerformanceTest.cpp

```diff
@@ -0,0 +1,423 @@
+      }
+    } catch (std::exception& ex) {
+      throw std::runtime_error(std::string(ex.what()) + " comparing to " + expected_result);
+    }
+  }
+
+  template <typename TSource>
+  static std::string to_json_object(
+      TSource const& src,
+      std::function<std::string(typename TSource::const_reference)> extractName,
+      std::function<std::string(typename TSource::const_reference)> extractValue) {
+    bool is_first{true};
+    return "{" +
+           std::accumulate(src.begin(), src.end(), std::string(),
+                           [&is_first, extractName,
+                            extractValue](std::string partial, auto const& element) {
+                             if (is_first) {
+                               is_first = false;
+                             } else {
+                               partial += ",";
+                             }
+                             partial += "\"" + extractName(element) +
+                                        "\": " + extractValue(element);
+                             return partial;
+                           }) +
+           "}";
+  }
+
+  template <typename TSource>
+  static std::string to_json_object(TSource const& src) {
+    return to_json_object(
+        src, [](auto const& element) { return element.first; },
+        [](auto const& element) { return element.second; });
+  }
+};
+
+// hammer a single key deep in the hierarchy 100000 times or so.
+//  We test long paths.
+//  Performance relative to depth.
+
+TEST_F(StorePerformanceTest, single_deep_key_writes) {
+  std::map<std::string, std::string> obj{};
+  for (int i{}; i < hammer_times; ++i) {
+    obj["a/b/c/d/e/f/g/h"] = std::to_string(i);
+    write(std::vector<std::vector<std::string>>{{to_json_object(obj)}});
+  }
+}
+
+TEST_F(StorePerformanceTest, single_deep_key_writes_and_reads) {
+  std::map<std::string, std::string> obj{};
+  auto const key {"a/b/c/d/e/f/g/h"};
+  auto const json_read_key {std::string("[[\"") + key + "\"]]"};
+  for (int i{}; i < hammer_times; ++i) {
+    obj[key] = std::to_string(i);
+    write(std::vector<std::vector<std::string>>{{to_json_object(obj)}});
+    auto result {read(json_read_key)};
+  }
+}
+
+TEST_F(StorePerformanceTest, single_deep_key_reads) {
+  std::map<std::string, std::string> obj{};
+  auto const key {"a/b/c/d/e/f/g/h"};
+  write(std::vector<std::vector<std::string>>{{to_json_object(obj)}});
+  auto const json_read_key {std::string("[[\"") + key + "\"]]"};
+  for (int i{}; i < hammer_times; ++i) {
+    obj[key] = std::to_string(i);
+    auto result {read(json_read_key)};
+  }
+}
+
+// write lots of different keys in different places
+//  Random-access performance
+
+static std::string rand_path(const size_t len = 10u) {
+  static auto const alphabet {"abcdefghijklmnopqrstuvwxyz0123456789"};
+  std::string result;
+  std::generate_n(std::back_inserter(result), 1 + std::rand() % (len-1), [](){
```

Comment:
From cppreference: "rand() is not recommended for serious random-number generation needs. It is recommended to use C++11's random number generation facilities to replace rand(). (since C++11)"

Reply:
And: I would generate the keys first and store them beforehand such that the actual measurement only measures the Store operations themselves.

Reply:
About the first comment, actually this was a conscious option. In this case I want to easily obtain the same items time and again, which makes rand() ideal (we only leave the default seed or force it with srand, and we get the exact same pseudorandomness). "Serious" in terms of randomness is not what I am after here.

Reply:
To the second comment, of course; actually if you look at some of the tests that is the approach I am taking.